### DISCUSSION TITLE
如果有插件在on_startup过程中报错，导致nb2退出，如何安全的进行清理工作

### DISCUSSION THREAD

**Discussion by shoucandanghehe** [nonebot/discussions#312] [问答] [answered]
我的插件在on_startup的时候建立了数据库连接
但是如果有其他on_startup函数在启动的时候抛出错误（比如依赖检查失败抛ImportError
数据库连接就没办法用on_shutdown关掉了，会导致bot卡死 只能kill
一个模拟例子（↓
![image](https://user-images.githubusercontent.com/51957264/216765361-0e365d09-44e1-4235-b5fe-e6bf6d23589f.png)

**Comment by yanyongyu** [comment, marked as answer]
uvicorn在startup失败时直接退出，不会执行shutdown。

https://github.com/encode/uvicorn/blob/master/uvicorn/server.py#L88-L91

**Reply by shoucandanghehe** [reply]
所以说需要一个方法（

**Reply by Lancercmd** [reply]
如果属于易故障的行为，你可以在 `on_bot_connect` 的时候才建立数据库连接和执行初始化任务

**Reply by shoucandanghehe** [reply]
> 如果属于易故障的行为，你可以在 `on_bot_connect` 的时候才建立数据库连接和执行初始化任务

可以是可以，但是又要避免重复运行，稍微有点麻烦（

**Reply by yanyongyu** [reply]
一般来说数据库启动连接池的时候并不会直接进行连接，照理来说退出是不会卡住的，你用的什么orm

**Reply by shoucandanghehe** [reply]
tortoise orm

**Reply by shoucandanghehe** [reply]
https://github.com/shoucandanghehe/nonebot-plugin-tetris-stats/blob/6c81cd56a0775a26fb8bf752ef30e159e9cf16b8/nonebot_plugin_tetris_stats/db/database.py#L37-L59

**Reply by yanyongyu** [reply]
你可以参考这个https://github.com/cscs181/QQ-GitHub-Bot/blob/master/src/providers/tortoise/__init__.py#L43-L50 ，一般来说初始化模型是在migrate中完成的，例如这个插件https://github.com/he0119/nonebot-plugin-datastore

**Reply by shoucandanghehe** [reply]
好 我明天看看） 第一次用orm 不是很熟

**Reply by yanyongyu** [reply]
你可以考虑直接用datastore插件试试（uy佬打钱@he0119
